oeps/oep-0050-hooks-extension-framework.rst

```diff
@@ -0,0 +1,235 @@
+
+Additionally, a directory in the same location will include individual files
+that cover each available trigger.
+
+The signature of the functions (input/output) of every trigger should be
+maintained in a stable way. The same applies for additions or removals from the
+list of public triggers, which should follow a path of deprecation first and
+preferably be given an optional hook where the same context information or code
+execution path can be accessed.
+
+The mechanism by which a trigger is connected to an action/filter will be a
+configurable python wrapper also contained in the platform core and by design
+must consider that:
+
+* Most actions will be implemented in plugins.
+* The order of execution for multiple actions/filters listening on a trigger
+  must be respected, and the results of previously executed filters/actions
+  should be made available to the current filter/action.
+* Actions should be easy to execute in an asynchronous way.
+* The results of executing actions or filters in an asynchronous way will be
+  ignored by the calling process. The will not modify the flow of the application
+
+
+The wrapper should present functions to register the trigger call, to call the
+configured actions as a pipeline, and to defer calling the functions as async
+tasks when configured. This wrapper should make the triggers, especially those
+that do not allow for application flow change, to be compatible with the Async
+Event Messaging described in OEP-41.
+In terms of error handling, it should allow the configuration of silent failures
+letting operators know the source of the error.
+
+
+Rationale
+=========
+
+Hooks are a way for one piece of code to interact/modify another piece of code
+at specific, pre-defined spots. Many of the business rules of an installation
+can be altered and extended if the right spots are allowed to be extended by
+open edX plugins. Initially, this proposal is meant as a way for community
+members to extend the platform, but it could even become a way for the
+installation at edx.org to implement its particular business rules in a way that
+is sufficiently separated from the Open edX core.
+
+This proposal draws heavy inspiration from WordPress, in which the same
+`two types of hooks`_ are defined. Leveraging the Python/Django based technology
+of open edX makes actions a great candidate to be executed asynchronously in a
+different process as they are meant to be called with some context data, act on
+it, and return silently to the trigger location.
+Filters on the other hand are designed so that they are given some context data,
+act on it, and return a modified version for the rest of the core-code to use in
+the next steps of execution. We intend to make the two kinds of hooks
+interchangeable such that developers are not limited by the framework but
+instead given maximum flexibility on its usage to attain their extension goals.
```

Comment:
[strengthening this proposal] It would be great if you could include a few Use Cases (feature-extension examples) for each of the 2 types of hooks. That would bring to light what each hook could enable for the platform.

Reply:
Added a few that I think illustrate what I have in mind.
https://github.com/edx/open-edx-proposals/pull/184/files#diff-f28146adea40eb552912482a8b24a19107252beba3e25295413bcd5ef8927628R196-R216